torch/ao/quantization/dynamic_tracing/README.md

```diff
@@ -0,0 +1,259 @@
+# Quantization with dynamic tracing (final name TBD)
+
+Note: this code is an early prototype and the API may change at any time.
+
+Quantization with dynamic tracing (final name TBD) is a prototype of automated
+quantization syntax transforms for PyTorch Eager mode. High level algorithm:
+
+1. take a user model and an example input
+2. trace the model with example input and record the subgraphs of seen quantizeable ops
+3. define quantization syntax transforms over the seen subgraphs
+4. during execution of user code, dynamically call into the subgraphs from (3) when necessary
+
+## User API overview
+
+```
+from torch.ao.quantization._quantize_dynamic_tracing import prepare, convert
+
+m = M(...)
+mp = prepare(m, example_input)
+# calibration (not shown)
+mq = convert(mp)
+```
+
+## Framework concepts overview
+
+The framework is defined with two major concepts:
+
+1. Each non-leaf module has a child of type `AutoQuantizationState`, stored
+under the `_auto_quant_state` attribute name. This child contains the quantization
+related state such as captured subgraphs and observers.
+2. During program execution, the framework overrides each module and quantizeable
+op to call hooks on the objects defined in (1), to implement the quantization
+syntax transforms.
+
+For example, imagine a user module such as this one:
+
+```
+class Child(torch.nn.Module):
+    def forward(self, x):
+        return x + x
+
+class Parent(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv = torch.nn.Conv2d(1, 1, 1)
+        self.child = Child()
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.child(x)
+        return x
+
+m = Parent().eval()
+```
+
+After model creation, the model hierarchy looks like
+
+```
+m (Parent)
+|- conv (torch.nn.Conv2d)
+|- child (Child)
+```
+
+After adding auto-observe, the model hierarchy looks like
+
+```
+m (Parent)
+|- conv (torch.nn.Conv2d)
+|- _auto_quant_state (AutoQuantizationState)
```

Comment:
should this be indented?

Reply:
this refers to `m._auto_quant_state`, `m.conv` does not have one because it is a leaf